drivers/platform/mbed/mbed_irq.cpp

```diff
@@ -0,0 +1,569 @@
+ *  - Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in
+ *    the documentation and/or other materials provided with the
+ *    distribution.
+ *  - Neither the name of Analog Devices, Inc. nor the names of its
+ *    contributors may be used to endorse or promote products derived
+ *    from this software without specific prior written permission.
+ *  - The use of this software may or may not infringe the patent rights
+ *    of one or more patent holders.  This license does not release you
+ *    from the requirement that you obtain separate licenses from these
+ *    patent holders to use this software.
+ *  - Use of the software either in source or binary form, must be run
+ *    on or directly connected to an Analog Devices Inc. component.
+ *
+ * THIS SOFTWARE IS PROVIDED BY ANALOG DEVICES "AS IS" AND ANY EXPRESS OR
+ * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, NON-INFRINGEMENT,
+ * MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
+ * IN NO EVENT SHALL ANALOG DEVICES BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, INTELLECTUAL PROPERTY RIGHTS, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+*******************************************************************************/
+
+/******************************************************************************/
+/***************************** Include Files **********************************/
+/******************************************************************************/
+
+#include <stdio.h>
+#include <mbed.h>
+
+// Platform support needs to be C-compatible to work with other drivers
+#ifdef __cplusplus
+extern "C"
+{
+#endif
+
+#include "no-os/error.h"
+#include "no-os/irq.h"
+#include "mbed_irq.h"
+#include "mbed_uart.h"
+
+using namespace std::chrono;
+
+/**
+* @struct mbed_irq_callback_desc.
+* @brief Structure holding the callback functions for mbed irqs.
+* @note  The callback functions are directly mapped with 'irq_id' enum.
+* Mbed doesn't provide a single callback function for all external/GPIO
+* IRQ events. A seperate callback function is required to be registered
+* for every IRQ instance. Also, the 'callback_function' required for
+* mbed layer expects void argument and void return type.
+*/
+typedef struct {
+	struct callback_desc callback_ext_int_id1;
+	struct callback_desc callback_ext_int_id2;
+	struct callback_desc callback_ext_int_id3;
+	struct callback_desc callback_ext_int_id4;
+	struct callback_desc callback_ext_int_id5;
```

Comment:
We could make those an array `struct callback_desc callback_ext_int_id[5];` and refer to them by index. I think this will help with the double `switch` scenario in the register function.

Reply:
Make sense. I have restructured this bit differently (using array of callback structure) to simplify some of the later code.